src/styles/Styled.js

```diff
@@ -0,0 +1,330 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import warning from 'warning';
+import getDisplayName from 'recompose/getDisplayName';
+import contextTypes from 'react-jss/lib/contextTypes';
+import { create } from 'jss';
+import jssGlobal from 'jss-global';
+import jssNested from 'jss-nested';
+import jssCamelCase from 'jss-camel-case';
+import jssDefaultUnit from 'jss-default-unit';
+import jssVendorPrefixer from 'jss-vendor-prefixer';
+import jssPropsSort from 'jss-props-sort';
+import * as ns from 'react-jss/lib/ns';
+import createMuiTheme from './createMuiTheme';
+import themeListener from './themeListener';
+import createGenerateClassName from './createGenerateClassName';
+import getStylesCreator from './getStylesCreator';
+
+export const preset = () => ({
+  plugins: [
+    jssGlobal(),
+    jssNested(),
+    jssCamelCase(),
+    jssDefaultUnit(),
+    jssVendorPrefixer(),
+    jssPropsSort(),
+  ],
+});
+
+// New JSS instance.
+const jss = create(preset());
+
+// Use a singleton or the provided one by the context.
+const generateClassName = createGenerateClassName();
+
+// Global index counter to preserve source order.
+// As we create the style sheet during componentWillMount lifecycle,
+// children are handled after the parents, so the order of style elements would
+// be parent->child. It is a problem though when a parent passes a className
+// which needs to override any childs styles. StyleSheet of the child has a higher
+// specificity, because of the source order.
+// So our solution is to render sheets them in the reverse order child->sheet, so
+// that parent has a higher specificity.
+let indexCounter = Number.MIN_SAFE_INTEGER;
+
+export const sheetsManager: Map<*, *> = new Map();
+
+// We use the same empty object to ref count the styles that don't need a theme object.
+const noopTheme = {};
+
+// In order to have self-supporting components, we rely on default theme when not provided.
+let defaultTheme;
+
+function getDefaultTheme() {
+  if (defaultTheme) {
+    return defaultTheme;
+  }
+
+  defaultTheme = createMuiTheme();
+  return defaultTheme;
+}
+
+class Styled extends React.Component {
+  static defaultProps = {
+    withTheme: false,
+  };
+  constructor(props, context) {
+    super(props, context);
+
+    const { muiThemeProviderOptions } = this.context;
+
+    this.jss = this.context[ns.jss] || jss;
+
+    if (muiThemeProviderOptions) {
+      if (muiThemeProviderOptions.sheetsManager) {
+        this.sheetsManager = muiThemeProviderOptions.sheetsManager;
+      }
+
+      this.disableStylesGeneration = muiThemeProviderOptions.disableStylesGeneration;
+    }
+
+    this.stylesCreator = getStylesCreator(props.styles);
+    if (this.stylesCreator.options.index === undefined) {
+      indexCounter += 1;
+      this.stylesCreator.options.index = indexCounter;
+    }
+
+    warning(
+      indexCounter < 0,
```

Comment:
< 0 ?

Reply:
This warning isn't that useful. It' in case you called the component many many times. But would have probably found the issue before.

Reply:
Still, the warning doesn't harm.